### PR TITLE
[Fix] raise execution errors for runnable tasks

### PR DESCRIPTION
**Pull request by MichelleArk**
resolves https://github.com/dbt-labs/dbt-core/issues/8166
[docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) dbt-labs/docs.getdbt.com/#  N/A

### Problem

Root cause is this `return` statement within a `finally` clause: https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/task/runnable.py#L386

From: https://docs.python.org/3/tutorial/errors.html#defining-clean-up-actions
> If the finally clause executes a [break](https://docs.python.org/3/reference/simple_stmts.html#break), [continue](https://docs.python.org/3/reference/simple_stmts.html#continue) or [return](https://docs.python.org/3/reference/simple_stmts.html#return) statement, exceptions are not re-raised.

The initial motivation for this change was to ensure that the current node as well as any other nodes involved in the command had their run result metadata returned for writing to run_results.json. https://github.com/dbt-labs/dbt-core/pull/8066/files#r1260138324

### Solution

- move the return statement out of the finally block, this is typically discouraged for reasons such as this issue. 
- return node_results on fail-fast to ensure the [initial fail-fast issue](https://github.com/dbt-labs/dbt-core/issues/7785) remains resolved (test coverage is here as well under tests/functional/fail_fast)
### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

🎩 
```sh
❯ dbt show --inline 'slect asdlkjfsld;j'
16:28:51  Running with dbt=1.7.0-a1
16:28:51  target not specified in profile 'postgres', using 'default'
16:28:51  Registered adapter: postgres=1.7.0-a1
16:28:51  Found 10 models, 3 seeds, 20 tests, 0 sources, 0 exposures, 0 metrics, 353 macros, 0 groups, 0 semantic models
16:28:51  
16:28:52  Concurrency: 1 threads (target='default')
16:28:52  
16:28:52  Encountered an error:
Runtime Error
  Database Error in sql_operation inline_query (from remote system.sql)
    syntax error at or near "slect"
    LINE 2: slect asdlkjfsld;j
```

### CODE REVIEW

File: tests/functional/dependencies/test_local_dependency.py [hunk 1]
```diff
@@ -208,9 +207,8 @@ def models(self):
 
     def test_missing_dependency(self, project):
         # dbt should raise a runtime exception
-        res = run_dbt(["compile"], expect_pass=False)
-        assert len(res) == 1
-        assert res[0].status == RunStatus.Error
```
Comment (MichelleArk) [inline tests/functional/dependencies/test_local_dependency.py:213]: reverts the changes made here: https://github.com/dbt-labs/dbt-core/pull/8066/files#diff-4f5f78d9b3b6f6eda81916de152b5c17b9d1ef14e40b7a2031c4d10f601d6474L210-L211

On closer inspection, I'm not sure why this test was modified, it is a change in behaviour for non fail-fast functionality.

File: core/dbt/task/runnable.py [hunk 2]
```diff
@@ -375,15 +375,17 @@ def execute_nodes(self):
...
         except KeyboardInterrupt:
             self._cancel_connections(pool)
             print_run_end_messages(self.node_results, keyboard_interrupt=True)
             raise
-        finally:
-            pool.close()
-            pool.join()
-            return self.node_results
+
+        pool.close()
```
Comment (ChenyuLInx) [inline core/dbt/task/runnable.py:385]: Question, do we want to always do `pool.close() , pool.join()` before join? If yes, based on the document that finally will be executed before `break, return, continue`, should we consider just write it as
```python
try:
  xxxx
  return  self.node_results
except:
  xxxx
finally:
  pool.close()
  pool.join()

**Review by ChenyuLInx — COMMENTED** [review]
One question regarding pool them it should be good to go

File: core/dbt/task/runnable.py (hunk 2 above)
Comment (MichelleArk) [reply core/dbt/task/runnable.py:385]: That may be a very sound thing to do, but I think the safest path forward to ensure the regression (and any others that may be lingering) are fixed is to go back to the state before this change: https://github.com/dbt-labs/dbt-core/pull/8066/files#diff-f01af78b7eaa1bacc301afa75eea4c7cdbe03064b3913bc2ce3f6bf01d0bc125R383-R386, where there was no `finally`

**Review by ChenyuLInx — APPROVED** [review]

**Comment by MichelleArk** [comment]
Additional 🎩 for expected fail fast behaviour: 

```sh 
❯ dbt run --select +downstream --fail-fast --project-dir ~/src/jaffle_shop
17:16:34  Running with dbt=1.7.0-a1
17:16:34  target not specified in profile 'postgres', using 'default'
17:16:34  Registered adapter: postgres=1.7.0-a1
17:16:34  Unable to do partial parsing because of a version mismatch
17:16:34  Found 10 models, 3 seeds, 20 tests, 0 sources, 0 exposures, 0 metrics, 353 macros, 0 groups, 0 semantic models
17:16:34  
17:16:35  Concurrency: 1 threads (target='default')
17:16:35  
17:16:35  1 of 2 START sql table model test_arky.upstream ................................ [RUN]
17:16:35  1 of 2 ERROR creating sql table model test_arky.upstream ....................... [ERROR in 0.05s]
17:16:35  CANCEL query model.jaffle_shop.upstream ........................................ [CANCEL]
17:16:35  
17:16:35  Database Error in model upstream (models/fail_fast/upstream.sql)
17:16:35    division by zero
17:16:35    compiled Code at target/run/jaffle_shop/models/fail_fast/upstream.sql
17:16:35  
17:16:35  Finished running 2 table models in 0 hours 0 minutes and 0.62 seconds (0.62s).
17:16:35  
17:16:35  Completed with 2 errors and 0 warnings:
17:16:35  
17:16:35  Database Error in model upstream (models/fail_fast/upstream.sql)
17:16:35    division by zero
17:16:35    compiled Code at target/run/jaffle_shop/models/fail_fast/upstream.sql
17:16:35  
17:16:35  Skipping due to fail_fast
17:16:35  
17:16:35  Done. PASS=0 WARN=0 ERROR=1 SKIP=1 TOTAL=2
```

Looking at run results, the skipped nodes are there as expected :) 

```json
 "results": [
        {
            "status": "error",
            "timing": [
                {
                    "name": "compile",
                    "started_at": "2023-07-28T17:16:35.229905Z",
                    "completed_at": "2023-07-28T17:16:35.231937Z"
                },
                {
                    "name": "execute",
                    "started_at": "2023-07-28T17:16:35.232805Z",
                    "completed_at": "2023-07-28T17:16:35.275653Z"
                }
            ],
            "thread_id": "Thread-1 (worker)",
            "execution_time": 0.05124211311340332,
            "adapter_response": {},
            "message": "Database Error in model upstream (models/fail_fast/upstream.sql)\n  division by zero\n  compiled Code at target/run/jaffle_shop/models/fail_fast/upstream.sql",
            "failures": null,
            "unique_id": "model.jaffle_shop.upstream"
        },
        {
            "status": "skipped",
            "timing": [],
            "thread_id": "MainThread",
            "execution_time": 0.0,
            "adapter_response": {},
            "message": "Skipping due to fail_fast",
            "failures": null,
            "unique_id": "model.jaffle_shop.downstream"
        }
    ],
```

**Pull request merged**